### PR TITLE
Generate specialization constant instructions

### PR DESCRIPTION
**Pull request by AlexeySotkin**
Translate SPIRV-friendly LLVM IR builtins represnting specialization constants `__spirv_SpecConstant` to corresponding SPIR-V instructions.

This approach works only for scalar constants, so support for `OpSpecConstantComposite` and `OpSpecConstantOp` is not implemented in this patch.

This PR also includes a patch to improve error handling for `llvm::getSpecConstInfo` API: Change return value from void to bool. Return `false` to indicate a failure. Return `true` on success.

### CODE REVIEW

File: lib/SPIRV/SPIRVWriter.cpp [hunk 1]
```diff
@@ -2039,6 +2042,28 @@ void LLVMToSPIRV::oclGetMutatedArgumentTypesByBuiltin(
...
 
+SPIRVValue *
+LLVMToSPIRV::transBuiltinToConstant(const std::string &DemangledName,
+                                    CallInst *CI) {
+  Op OC = getSPIRVFuncOC(DemangledName);
+  if(!isSpecConstantOpCode(OC))
+    return nullptr;
+  Type *Ty = CI->getArgOperand(1)->getType();
+  assert(Ty == CI->getType() && "Type mismatch!");
+  Value *V = CI->getArgOperand(1);
```
Comment (AlexeySachkov) [inline lib/SPIRV/SPIRVWriter.cpp]:
```c++
  Value *V = CI->getArgOperand(1);
  Type *Ty = V->getType();
  assert(Ty == CI->getType() && "Type mismatch!");
```

File: lib/SPIRV/SPIRVWriter.cpp [hunk 2]
```diff
@@ -2039,6 +2042,28 @@ void LLVMToSPIRV::oclGetMutatedArgumentTypesByBuiltin(
...
+  uint64_t Val = 0;
+  if(Ty->isIntegerTy())
```
Comment (AlexeySachkov) [inline lib/SPIRV/SPIRVWriter.cpp]:
```suggestion
  if (Ty->isIntegerTy())
```

The same at line 2057

File: lib/SPIRV/libSPIRV/SPIRVType.h [hunk 3]
```diff
@@ -191,7 +191,7 @@ class SPIRVTypeFloat : public SPIRVType {
     SPIRVCapVec CV;
     if (isTypeFloat(16)) {
       CV.push_back(CapabilityFloat16Buffer);
-      auto Extensions = getModule()->getExtension();
+      auto Extensions = getModule()->getSourceExtension();
```
Comment (AlexeySachkov) [inline lib/SPIRV/libSPIRV/SPIRVType.h]: Seems like unrelated change to me, is it possible to move it into a separate PR?

**Comment by AlexeySotkin** [comment]
@aratajew Please take a look.

**Comment by aratajew** [comment]
> @aratajew Please take a look.
Regarding to the second commit (c2e6c6c9): LGTM, thanks for implementing this.

File: lib/SPIRV/libSPIRV/SPIRVType.h (hunk 3 above)
Comment (AlexeySotkin) [reply lib/SPIRV/libSPIRV/SPIRVType.h]: What about moving it to a separate commit?

File: lib/SPIRV/SPIRVWriter.cpp (hunk 1 above)
Comment (AlexeySotkin) [reply lib/SPIRV/SPIRVWriter.cpp]: Fixed, thanks

File: lib/SPIRV/SPIRVWriter.cpp (hunk 2 above)
Comment (AlexeySotkin) [reply lib/SPIRV/SPIRVWriter.cpp]: Fixed, thanks.

File: lib/SPIRV/SPIRVReader.cpp [hunk 4]
```diff
@@ -3693,4 +3693,7 @@ void llvm::getSpecConstInfo(std::istream &IS,
       D.ignoreInstruction();
     }
   }
+  if (IS.fail())
+    return false;
+  return true;
```
Comment (AlexeySachkov) [inline lib/SPIRV/SPIRVReader.cpp]:
```c++
return !IS.fail();
```
